File: core/src/main/java/org/dreamexposure/discal/core/filter/GatewayEventFilter.java
package org.dreamexposure.discal.core.filter;

import org.slf4j.Marker;

import java.util.Arrays;
import java.util.List;

import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.Logger;
import ch.qos.logback.classic.turbo.TurboFilter;
import ch.qos.logback.core.spi.FilterReply;
import discord4j.discordjson.json.gateway.Opcode;
import discord4j.gateway.json.GatewayPayload;

@SuppressWarnings({"unused", "MethodWithMoreThanThreeNegations", "OverlyNestedMethod"})
public class GatewayEventFilter extends TurboFilter {

    private String include;
    private String exclude;
    private List<String> includedEvents;
    private List<String> excludedEvents;

    @Override
    public FilterReply decide(final Marker marker, final Logger logger, final Level level, final String format,
                              final Object[] params, final Throwable t) {
        if (params != null && logger.getName().startsWith("discord4j.gateway.inbound")) {
            for (final Object param : params) {
                if (param instanceof GatewayPayload) {
                    final GatewayPayload<?> payload = (GatewayPayload<?>) param;
                    if (Opcode.DISPATCH.equals(payload.getOp())) {
                        if (this.excludedEvents != null) {
                            if (this.excludedEvents.contains(payload.getType())) {
                                return FilterReply.DENY;
                            }
                        } else if (this.includedEvents != null) {
                            if (!this.includedEvents.contains(payload.getType())) {
                                return FilterReply.DENY;
                            }
                        }
                    }
                }
            }
        }
        return FilterReply.NEUTRAL;
    }

    public void setInclude(final String include) {
        this.include = include;
    }

    public void setExclude(final String exclude) {
        this.exclude = exclude;
    }

    @Override
    public void start() {
        if (this.exclude != null && !this.exclude.trim().isEmpty()) {
            this.excludedEvents = Arrays.asList(this.exclude.split("[;,]"));
            super.start();
        } else if (this.include != null && !this.include.trim().isEmpty()) {
            this.includedEvents = Arrays.asList(this.include.split("[;,]"));
            super.start();
        }
    }
}

File: core/src/main/kotlin/org/dreamexposure/discal/core/filter/GatewayEventFilter.kt
package org.dreamexposure.discal.core.filter

import ch.qos.logback.classic.Level
import ch.qos.logback.classic.Logger
import ch.qos.logback.classic.turbo.TurboFilter
import ch.qos.logback.core.spi.FilterReply
import discord4j.discordjson.json.gateway.Opcode
import discord4j.gateway.json.GatewayPayload
import org.slf4j.Marker

@Suppress("unused")
class GatewayEventFilter : TurboFilter() {
    private var include: String? = null
    private var exclude: String? = null
    private var includedEvents: List<String>? = null
    private var excludedEvents: List<String>? = null
    override fun decide(marker: Marker?, logger: Logger?, level: Level?, format: String?,
                        params: Array<Any>?, t: Throwable?): FilterReply {
        if (params != null && logger!!.name.startsWith("discord4j.gateway.inbound")) {
            for (param in params) {
                if (param is GatewayPayload<*>) {
                    if (Opcode.DISPATCH == param.op) {
                        if (excludedEvents != null) {
                            if (excludedEvents!!.contains(param.type!!)) {
                                return FilterReply.DENY
                            }
                        } else if (includedEvents != null) {
                            if (!includedEvents!!.contains(param.type!!)) {
                                return FilterReply.DENY
                            }
                        }
                    }
                }
            }
        }
        return FilterReply.NEUTRAL
    }

    fun setInclude(include: String?) {
        this.include = include
    }

    fun setExclude(exclude: String?) {
        this.exclude = exclude
    }

    override fun start() {
        if (exclude != null && exclude!!.trim { it <= ' ' }.isNotEmpty()) {
            excludedEvents = mutableListOf(*exclude!!.split("[;,]").toTypedArray())
            super.start()
        } else if (include != null && include!!.trim { it <= ' ' }.isNotEmpty()) {
            includedEvents = mutableListOf(*include!!.split("[;,]").toTypedArray())
            super.start()
        }
    }
}
